src/NodeApi/Interop/JSSynchronizationContext.cs

```diff
@@ -42,6 +43,148 @@ public virtual void Dispose()
     public abstract void OpenAsyncScope();
 
     public abstract void CloseAsyncScope();
+
+    public void Post(Action action, bool allowSync = false)
+    {
+        if (allowSync && Current == this)
+        {
+            action();
+        }
+        else
+        {
+            Post(_ =>
+            {
+                if (IsDisposed) return;
+                action();
+            }, null);
+        }
+    }
+
+    public void Post(Func<Task> asyncAction, bool allowSync = false)
+    {
+        if (allowSync && Current == this)
+        {
+            _ = asyncAction();
+        }
+        else
+        {
+            Post((_) =>
+            {
+                if (IsDisposed) return;
+                _ = asyncAction();
+            }, null);
+        }
+    }
+
+    public void Run(Action action)
+    {
+        if (Current == this)
+        {
+            action();
+        }
+        else
+        {
+            Exception? exception = null;
+            Send((_) =>
+            {
+                if (IsDisposed) return;
+                try
+                {
+                    action();
+                }
+                catch (Exception ex)
+                {
+                    exception = ex;
+                }
+            }, null);
+            if (exception != null)
+            {
+                throw new JSException("Exception thrown from JS thread.", exception);
+            }
+        }
+    }
+
+    public T Run<T>(Func<T> action)
+    {
+        if (Current == this)
+        {
+            return action();
+        }
+        else
+        {
+            T result = default!;
+            Exception? exception = null;
+            Send((_) =>
+            {
+                if (IsDisposed) return;
+                try
+                {
+                    result = action();
+                }
+                catch (Exception ex)
+                {
+                    exception = ex;
+                }
+            }, null);
+            if (exception != null)
+            {
+                throw new JSException("Exception thrown from JS thread.", exception);
+            }
+            return result;
+        }
+    }
+
+    public Task RunAsync(Func<Task> asyncAction)
+    {
+        if (Current == this)
+        {
+            return asyncAction();
+        }
+        else
+        {
+            TaskCompletionSource<bool> completion = new();
+            Send(async (_) =>
```

Comment:
>Send [](http://example.com/codeflow?start=12&length=4)

It is better to avoid synchronous call for async invocation.

Reply:
Yes, this should be `Post`. Also in the other `RunAsync` overload below.